### ISSUE TITLE
Validating objects with unspecified properties

### ISSUE THREAD

**Issue by doomsower** [ianstormtaylor/superstruct#117]
Let's say I have a struct: `struct.object({foo: 'string'})`  
Then I pass an object: `s.validate({ foo: 'bar', bar: 'foo' })`  
Currently I am receiving an error with path `[ 'bar' ]` and message `Expected a value of type `undefined` for `bar` but received `"foo"`.`

I think that this behavior is incorrect, and the `path` should be empty array, i.e. the error belongs to whole struct and not to unspecified property. And the message should be adjusted accordingly.

The following use case illustrates why I think so. I face this situation quite often.  There is a form that receives data from backend, user edits the data, then submits it back. The thing is that what backend sends, what backend expects to receive, and what the form internally uses sometimes have different shapes (the form needs some extra data to look nice, or rich text needs to be transformed from markdown to internal representation, etc...). This leads to unspecified properties, and results in validation error. UI then checks paths of errors and displays them on corresponding UI controls. Those errors with empty path array (like multi-field errors) are then displayed in an ugly red box for whole form errors. The problem with current error paths is that there is no way I can know that unspecified fields errors belong to whole form, they look like they belong to particular UI control, but there is no such control on the form.

**Comment by doomsower** [comment]
It seems that it's a one line fix [here](https://github.com/ianstormtaylor/superstruct/blob/5d537e1017dc27faed6ecd9fbb42aa101850d0f0/src/kinds.js#L550), but I understand that it's a) breaking and b) opinionated change.

**Comment by doowb** [comment]
I think you can use [interface](https://github.com/ianstormtaylor/superstruct/blob/master/docs/reference.md#interface) to achieve your use case. This will only assert the properties in the interface, while [partial](https://github.com/ianstormtaylor/superstruct/blob/master/docs/reference.md#partial) will do the same, but only return the properties in the partial.

**Comment by doomsower** [comment]
@doowb ,  `interface` and `partial` do not produce errors on unspecified fields. And I need errors, not sanitized objects. Just double-checked it in node console.

<details><summary>node REPL</summary>
<p>

```js
> const { struct } = require('superstruct')
undefined
> const o = { foo: 'bar', bar: 'foo' }
undefined
> const schema = { foo: 'string' }
undefined
> const inter = struct.interface(schema)
undefined
> const part = struct.partial(schema)
undefined
> const obj = struct.object(schema)
undefined
> inter.validate(o)
[ undefined, { foo: 'bar', bar: 'foo' } ]
> part.validate(o)
[ undefined, { foo: 'bar' } ]
> obj.validate(o)
[ { TypeError: Expected a value of type `undefined` for `bar` but received `"foo"`.
    at Function.Struct.validate.value [as validate] (/Volumes/Projects/whitewater/packages/web/node_modules/superstruct/lib/index.js:1196:17)
    at repl:1:5
    at ContextifyScript.Script.runInThisContext (vm.js:50:33)
    at REPLServer.defaultEval (repl.js:240:29)
    at bound (domain.js:301:14)
    at REPLServer.runBound [as eval] (domain.js:314:12)
    at REPLServer.onLine (repl.js:468:10)
    at emitOne (events.js:121:20)
    at REPLServer.emit (events.js:211:7)
    at REPLServer.Interface._onLine (readline.js:280:10)
    data: { foo: 'bar', bar: 'foo' },
    path: [ 'bar' ],
    value: 'foo',
    reason: undefined,
    type: undefined,
    errors: [ [Object] ] } ]
```

</p>
</details>

**Comment by doowb** [comment]
Oh, I see now. I misunderstood your goal.

**Comment by ianstormtaylor** [comment]
@doomsower I feel your pain here. I think you can use the presence of `type` to determine whether the field is a "known" field with an expected type or not? 

Otherwise, I fear that people will have the same problem in the opposite direction, of not being able to pinpoint which subfield is the one that caused the error?

**Comment by doomsower** [comment]
Thanks, @ianstormtaylor 

Indeed I can distinguish unspecified fields based on `type`, this solves it for me.

But while I was toying with the node console I found something that I don't understand.
This falls into the same problem of mine - distinguish field-level errors from form-level errors and properly display them in UI.
```js
> const { struct } = require('superstruct')
undefined
> const schema1 = struct.object({ foo: 'string', name: 'string?' })
undefined
> const schema2 = struct.function((v) => v.foo === 'bar' ? v.name === 'foo' : true)
undefined
> const schema = struct.intersection([schema1, schema2])
undefined
> const o = { foo: 'bar', bar: 'foo', name: 1 }
undefined
> const [e] = schema.validate(o)
undefined
> e.errors
[ { data: { foo: 'bar', bar: 'foo', name: 1 },
    path: [ 'bar' ],
    value: 'foo',
    errors: [Circular],
    type: '{foo,name} & <function>' },
  { type: 'string | undefined',
    value: 1,
    data: { foo: 'bar', bar: 'foo', name: 1 },
    path: [ 'name' ],
    reason: null } ]

```
So here I have two field-level rules and one form-level rule, and I intersect them. Here I expected to get 3 errors:
  1. [x] name is number instead of `string?` at path `name`
  2. [ ] error with `type === undefined` at path `bar`, as you suggest and as in my original example
  3. [ ] some form-level error with empty path (for "if bar the foo rule")

Instead I've got one expected error (nr. 1) and one totally weird. So my criteria for form-level error `path is empty or type is undefined` is not working here. Looks like the type of nr.2 got masked by intersection. How do I solve this?

**Comment by ianstormtaylor** [comment]
@doomsower not exactly sure, the example is hard to follow. But it sounds like it may be a bug in `intersection` if what you're saying is right. If it is I'd accept a PR with an explanation of where it comes from.

**Comment by ianstormtaylor** [comment]
@doomsower can you paste the console output there?

> It just takes first invalid operand and replaces its root error with itself. In the example above, first operand is object. It contains two errors with name and password paths. Those two errors are really at the same level and we arbitrary make one of them root error and then place all errors into errors property. Then intersection would overwrite one of two really equally-weighted errors but leave the other as it is.

Do you have a suggestion for which error to take? I don't think it's "wrong" really for whatever the logic is. The "root" error is for people who only ever want one error surfaced, which is the simpler common case, and in that case they're all equivalent IMO. Taking the first surfaced error seems reasonable to me.

> Also it stops validation after first invalid operand. From practical point of view I think this is wrong. object doesn't stop validation after first invalid field.

This makes sense to me, I'd be okay with it not stopping here for `intersection` I think.

**Comment by toxsick** [comment]
Hey guys, does my #118 help here perhaps?

**Comment by tymfear** [comment]
So, any resolution on this? I wanted to use this library in my integration tests, but it returns only first error, and, of course, I'd like to see in my test report all the errors at once and re-running tests again and again if more that one field does not match the schema.
